### PR TITLE
Add an entity to lists from its profile page

### LINKED ISSUE #112: create the ability to add an entity to a list from it's profile page
On the entity sidebar under lists add a button that creates a popup that allow users to add the entity to already existing lists

### PR DESCRIPTION
**Pull request by robjlucas**
Resolves #112 

Screenshot from 2020-05-21 14-22-05
Screenshot from 2020-05-21 14-22-35

### CODE REVIEW

**Comment by robjlucas** [comment]
@aepyornis this will presumably require list searches and controller or feature tests, but I've so far been unable to test list searching in RSpec. Passing a `q` param to the  `search_controller#index` leads to MySQL connection errors. This is inside the docker container. Any ideas?

`list_ids = List.search(
         Riddle::Query.escape(params[:q]),
         with: { is_deleted: 0, is_admin: is_admin }
       ).map(&:id)
     
     ThinkingSphinx::ConnectionError:
       Error connecting to Sphinx via the MySQL protocol. Can't connect to MySQL server on '127.0.0.1' (115)`

File: spec/controllers/lists_controller_spec.rb [hunk 1]
```diff
@@ -44,6 +44,42 @@
...
+        end
+      end
+    end
+
+    after(:all) do
+      teardown_sphinx {  }
+    end
+
+    it 'returns search results in HTML' do
+      get :index, params: { q: 'interesting' }
```
Comment (robjlucas) [inline spec/controllers/lists_controller_spec.rb:68]: @aepyornis some difficulties testing stuff that involves Sphinx. It seems I have to run `RAILS_ENV=test bin/rails ts:rebuild` every time I want to run a test involving Sphinx. It then spends ages on `Generating configuration to /littlesis/config/test.sphinx.conf`. Any suggestions?

```
ThinkingSphinx::SphinxError:
       unknown local index(es) 'list_core' in search request - SELECT * FROM `list_core`, `list_delta` WHERE MATCH('interesting') AND `is_deleted` = 0 AND `is_admin` = 0 AND `sphinx_deleted` = 0 LIMIT 0, 20; SHOW META
     # /usr/local/bundle/gems/thinking-sphinx-4.4.1/lib/thinking_sphinx/connection/client.rb:65:in `rescue in perform'
     # /usr/local/bundle/gems/thinking-sphinx-4.4.1/lib/thinking_sphinx/connection/client.rb:61:in `perform'
     # /usr/local/bundle/gems/thinking-sphinx-4.4.1/lib/thinking_sphinx/connection/client.rb:51:in `check_and_perform'
     # /usr/local/bundle/gems/thinking-sphinx-4.4.1/lib/thinking_sphinx/connection/client.rb:36:in `query_all'
     # /usr/local/bundle/gems/thinking-sphinx-4.4.1/lib/thinking_sphinx/search/batch_inquirer.rb:19:in `block in results'
     # /usr/local/bundle/gems/thinking-sphinx-4.4.1/lib/thinking_sphinx/connection.rb:38:in `block in take'
     # /usr/local/bundle/gems/innertube-1.1.0/lib/innertube.rb:138:in `take'
     # /usr/local/bundle/gems/thinking-sphinx-4.4.1/lib/thinking_sphinx/connection.rb:36:in `take'
     # /usr/local/bundle/gems/thinking-sphinx-4.4.1/lib/thinking_sphinx/search/batch_inquirer.rb:18:in `results'
     # /usr/local/bundle/gems/thinking-sphinx-4.4.1/lib/thinking_sphinx/middlewares/inquirer.rb:11:in `block in call'
     # /usr/local/bundle/gems/thinking-sphinx-4.4.1/lib/thinking_sphinx/logger.rb:5:in `log'
     # /usr/local/bundle/gems/thinking-sphinx-4.4.1/lib/thinking_sphinx/middlewares/inquirer.rb:10:in `call'
     # /usr/local/bundle/gems/thinking-sphinx-4.4.1/lib/thinking_sphinx/middlewares/geographer.rb:13:in `call'
     # /usr/local/bundle/gems/thinking-sphinx-4.4.1/lib/thinking_sphinx/middlewares/sphinxql.rb:16:in `call'
     # /usr/local/bundle/gems/thinking-sphinx-4.4.1/lib/thinking_sphinx/middlewares/valid_options.rb:9:in `call'
     # /usr/local/bundle/gems/thinking-sphinx-4.4.1/lib/thinking_sphinx/middlewares/stale_id_filter.rb:12:in `call'
     # /usr/local/bundle/gems/middleware-0.1.0/lib/middleware/runner.rb:31:in `call'
     # /usr/local/bundle/gems/middleware-0.1.0/lib/middleware/builder.rb:102:in `call'
     # /usr/local/bundle/gems/thinking-sphinx-4.4.1/lib/thinking_sphinx/search.rb:95:in `populate'
     # /usr/local/bundle/gems/thinking-sphinx-4.4.1/lib/thinking_sphinx/search.rb:145:in `method_missing'
     # ./app/controllers/lists_controller.rb:59:in `index'
     # /usr/local/bundle/gems/rails-controller-testing-1.0.4/lib/rails/controller/testing/template_assertions.rb:61:in `process'
     # /usr/local/bundle/gems/devise-4.7.1/lib/devise/test/controller_helpers.rb:35:in `block in process'
     # /usr/local/bundle/gems/devise-4.7.1/lib/devise/test/controller_helpers.rb:102:in `catch'
     # /usr/local/bundle/gems/devise-4.7.1/lib/devise/test/controller_helpers.rb:102:in `_catch_warden'
     # /usr/local/bundle/gems/devise-4.7.1/lib/devise/test/controller_helpers.rb:35:in `process'
     # /usr/local/bundle/gems/rails-controller-testing-1.0.4/lib/rails/controller/testing/integration.rb:13:in `block (2 levels) in <module:Integration>'
     # ./spec/controllers/lists_controller_spec.rb:68:in `block (3 levels) in <main>'
     # /usr/local/bundle/gems/bootsnap-1.4.6/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:55:in `load'
     # /usr/local/bundle/gems/bootsnap-1.4.6/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:55:in `load'
     # /usr/local/bundle/gems/spring-commands-rspec-1.0.4/lib/spring/commands/rspec.rb:18:in `call'
     # -e:1:in `<main>'
     # ------------------
     # --- Caused by: ---
     # Mysql2::Error:
     #   unknown local index(es) 'list_core' in search request
     #   /usr/local/bundle/gems/mysql2-0.5.3/lib/mysql2/client.rb:131:in `_query'
```

Comment (aepyornis) [reply spec/controllers/lists_controller_spec.rb:68]: Interesting. Are you running those tests outside docker?

While most tests run okay outside docker, the sphinx tests has never worked unless I run this _inside_ docker (`littlesis test spec/controllers/lists_controller_spec.rb  `)

But I've not seen that error before or at least not in some time.

Comment (robjlucas) [reply spec/controllers/lists_controller_spec.rb:68]: Thanks @aepyornis. I've tried both inside and outside the docker container (including installing Sphinx etc on my dev machine), and get similar results in both. Do you rebuild the indexes every time you want to run a test that involves Sphinx?

**Comment by robjlucas** [comment]
> This is great. Thank you! There's a few changes that needs to be sorted out -- not very oblivious things, so I ought to have alerted you to these pitfalls before. Sorry about this confusing permissions mess I'm about to spring on you. Feel free to say I want to move on from this nonsense and I'll help out...
> 
> Lists have a slightly more complicated permission structure. Each list can have three states:
> Open, Closed, and Private.
> 
> Private lists can only be viewed or updated by the user who created it.
> Closed lists are publicly viewable but can only be updated by the user who created it
> Open lists are publicly viewable and can be updated by anyone with the "list" ability.
> 
> The logic for this is in app/utility/permissions:
> 
> https://github.com/public-accountability/littlesis-rails/blob/712ad228049eb7f0e805b98df1fb79ff438fb7db/app/utility/permissions.rb#L90-L95
> 
> I like that you've adapted the /lists route to return json which is a totally sensible thing to do. However, that endpoint will include Closed lists can cannot updated by all users.
> 
> For users WITH the list ability it should return:
> 
>     * all lists they've created (any permission state)
> 
>     * Any open list
> 
> 
> For Users WITHOUT the list ability it should only return lists they've created.
> 
> This one is a quick change: in `app/views/entities/_sidebar.html.erb`, `form_for` should go outside of the cache here --
> 
> https://github.com/public-accountability/littlesis-rails/blob/712ad228049eb7f0e805b98df1fb79ff438fb7db/app/views/entities/_sidebar.html.erb#L65
> -- because caches are shared among all users. We should also surround the form with `if user_signed_in?` as happens [here](https://github.com/public-accountability/littlesis-rails/blob/712ad228049eb7f0e805b98df1fb79ff438fb7db/app/views/entities/_sidebar.html.erb#L55) since anonymous users cannot make changes.

Thanks for the very clear explanation @aepyornis . Note to self for future: pay more attention to how the actual site works rather than trying to figure out everything from looking at code. Which is to say, I was somewhat aware of private lists etc, so might have figured this out if I had thought it through rather than diving into code.

Anyway, I've put in pretty exhaustive permission-checking now, if you don't mind taking another look.

**Comment by robjlucas** [comment]
Actually @aepyornis it just occurred to me that we also should do some filtering here, so that closed lists don't appear in the results, so this is still not quite there, but almost: https://github.com/public-accountability/littlesis-rails/blob/cf31304290f0b416f74aa948950b8cc6eb49462b/app/controllers/lists_controller.rb#L44

**Pull request merged**